Tests/Functional/Asynchronous/Expectations/main.swift

```diff
@@ -77,6 +79,414 @@ class ExpectationsTestCase: XCTestCase {
         waitForExpectations(timeout: -1.0)
     }
 
+    // PRAGMA MARK: - Multiple Expectations
+
+// CHECK: Test Case 'ExpectationsTestCase.test_multipleExpectations' started at \d+-\d+-\d+ \d+:\d+:\d+\.\d+
+// CHECK: Test Case 'ExpectationsTestCase.test_multipleExpectations' passed \(\d+\.\d+ seconds\)
+    func test_multipleExpectations() {
+        let foo = expectation(description: "foo")
+        let bar = XCTestExpectation(description: "bar")
+        DispatchQueue.global(qos: .default).asyncAfter(wallDeadline: .now() + 0.01) {
+            bar.fulfill()
+        }
+        DispatchQueue.global(qos: .default).asyncAfter(wallDeadline: .now() + 0.01) {
+            foo.fulfill()
+        }
+
+        XCTWaiter(delegate: self).wait(for: [foo, bar], timeout: 1)
+    }
+
+// CHECK: Test Case 'ExpectationsTestCase.test_multipleExpectationsEnforceOrderingCorrect' started at \d+-\d+-\d+ \d+:\d+:\d+\.\d+
+// CHECK: Test Case 'ExpectationsTestCase.test_multipleExpectationsEnforceOrderingCorrect' passed \(\d+\.\d+ seconds\)
+    func test_multipleExpectationsEnforceOrderingCorrect() {
+        let foo = expectation(description: "foo")
+        let bar = XCTestExpectation(description: "bar")
+        DispatchQueue.global(qos: .default).asyncAfter(wallDeadline: .now() + 0.01) {
+            foo.fulfill()
+            bar.fulfill()
+        }
+
+        XCTWaiter(delegate: self).wait(for: [foo, bar], timeout: 1, enforceOrder: true)
+    }
+
+// CHECK: Test Case 'ExpectationsTestCase.test_multipleExpectationsEnforceOrderingCorrectBeforeWait' started at \d+-\d+-\d+ \d+:\d+:\d+\.\d+
+// CHECK: Test Case 'ExpectationsTestCase.test_multipleExpectationsEnforceOrderingCorrectBeforeWait' passed \(\d+\.\d+ seconds\)
+    func test_multipleExpectationsEnforceOrderingCorrectBeforeWait() {
+        let foo = expectation(description: "foo")
+        let bar = XCTestExpectation(description: "bar")
+        foo.fulfill()
+        bar.fulfill()
+        XCTWaiter(delegate: self).wait(for: [foo, bar], timeout: 1, enforceOrder: true)
+    }
+
+// CHECK: Test Case 'ExpectationsTestCase.test_multipleExpectationsEnforceOrderingIncorrect' started at \d+-\d+-\d+ \d+:\d+:\d+\.\d+
+// CHECK: .*/Tests/Functional/Asynchronous/Expectations/main.swift:[[@LINE+6]]: error: ExpectationsTestCase.test_multipleExpectationsEnforceOrderingIncorrect : Failed due to expectation fulfilled in incorrect order: requires 'foo', actually fulfilled 'bar'
+// CHECK: Test Case 'ExpectationsTestCase.test_multipleExpectationsEnforceOrderingIncorrect' failed \(\d+\.\d+ seconds\)
+    func test_multipleExpectationsEnforceOrderingIncorrect() {
+        let foo = expectation(description: "foo")
+        let bar = XCTestExpectation(description: "bar")
+        DispatchQueue.global(qos: .default).asyncAfter(wallDeadline: .now() + 0.01) {
+            bar.fulfill()
+            foo.fulfill()
+        }
+        XCTWaiter(delegate: self).wait(for: [foo, bar], timeout: 1, enforceOrder: true)
+    }
```

Comment:
These are great tests! Thanks for writing them.

I remember Apple XCTest had some interesting behavior where the first of these tests fails with "Failed due to expectations fulfilled in incorrect order", but the second passes:

```objc
- (void)test_encforcingTheOrderOfInvertedExpectations_inTheMiddleOfAList_createsAnImpossibleAssertion {
    XCTestExpectation *expectationOne = [self expectationWithDescription:@"one"];
    [expectationOne fulfill];

    XCTestExpectation *invertedExpectationTwo = [self expectationWithDescription:@"inverted two"];
    invertedExpectationTwo.inverted = YES;

    XCTestExpectation *expectationThree = [self expectationWithDescription:@"three"];
    [expectationThree fulfill]; // Fails here: "Failed due to expectation fulfilled in incorrect order"

    [self waitForExpectations:@[expectationOne, invertedExpectationTwo, expectationThree]
    	              timeout:0.1
    	         enforceOrder:YES];
}

- (void)test_encforcingTheOrderOfInvertedExpectations_atTheEndOfAList_passes {
    XCTestExpectation *expectationOne = [self expectationWithDescription:@"one"];
    [expectationOne fulfill];

    XCTestExpectation *expectationTwo = [self expectationWithDescription:@"two"];
    [expectationTwo fulfill];

    XCTestExpectation *invertedExpectationThree = [self expectationWithDescription:@"inverted three"];
    invertedExpectationThree.inverted = YES;

    [self waitForExpectations:@[expectationOne, expectationTwo, invertedExpectationThree]
    	              timeout:0.1
    	         enforceOrder:YES];
}
```

I'd be curious if Apple XCTest still behaves this way, and if so, maybe we should add a test that swift-corelibs-xctest does the same?

Reply:
Thanks @modocache!

I found the bug you're referring to (rdar://problem/31657785) and it was fixed in Xcode 9. I confirmed that both tests you pasted above now pass in recent versions of Xcode (using Apple XCTest). And that issue is now covered by the unit test I added in this PR named `test_combiningInverseAndStandardExpectationsWithOrderingEnforcement`.

Reply:
Awesome! Thanks for checking that.